### PR TITLE
units: run "bootctl update" automatically after boot, to ensure boot loader is regularly updated

### PR DESCRIPTION
**Pull request by poettering**

### CODE REVIEW

**Comment by poettering** [comment]
this is 240 material, of course.

**Comment by yuwata** [comment]
CentOS CI failed with the following:
```
TEST-01-BASIC: ● systemd-boot-update.service loaded failed failed Automatic Boot Loader Update
```

**Comment by bluca** [comment]
> this is **_240_** material, of course.

We'll need a flux capacitor then, and 1.21 gigawatt :astonished:

**Review by keszybz — COMMENTED** [review]
I think we should have had something like this from the start ;)

Nevertheless, I think this should be opt-in. As you say, it's "primarily intended from image-based update systems". Those are still a tiny minority. For the more traditional "package-based systems", this should be done through packaging. I think this is something to change e.g. in Fedora packaging, but the service should not be enabled there.

**Comment by poettering** [comment]
> Nevertheless, I think this should be opt-in. As you say, it's "primarily intended from image-based update systems". Those are still a tiny minority. For the more traditional "package-based systems", this should be done through packaging. I think this is something to change e.g. in Fedora packaging, but the service should not be enabled there.

It's now enabled via presets and via build scripts. Does this even matter for Fedora?

I'd really like to keep this enabled from upstream, since we should try hard to make this as graceful as possible, i.e. that this is a NOP in all expected cases it should be a NOP. I mean, if this wasn't enabled by default the CI wouldn't have caught that I fucked up in the first push of this PR.

I mean, I am fine if Fedora turns this off by default once this lands in Fedora, but upstream I think this should be on I think.

Note that the automatic update is designed to be a NOP if the boot loader is already up-to-date. That means in package-based distros it shouldn't hurt, it shouldn't even be noticable, but it might clean-up incomplete updates sooner or later. Which I think is at worst something that doesn't matter and at best adds a bit of robustness.

Also, isn't the sd-boot EFI binary its own RPM anyway on Fedora? If so this unit file should be shipped in that I figure.

**Comment by poettering** [comment]
I force pushed a new version btw, which hopefully fixes the CI issue

**Comment by poettering** [comment]
And yes, of course I mean *v250* material, not *v240* material ;-)

File: units/systemd-boot-update.service [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+Documentation=man:bootctl(1)
+DefaultDependencies=no
+Conflicts=shutdown.target
+After=local-fs.target
+Before=sysinit.target shutdown.target systemd-update-done.service
+
+[Service]
+Type=oneshot
+RemainAfterExit=yes
+ExecStart={{BINDIR}}/bootctl --no-variables --graceful update
```
Comment (mrc0mmand) [inline units/systemd-boot-update.service]:
```suggestion
ExecStart={{ROOTBINDIR}}/bootctl --no-variables --graceful update
```

I guess?

Comment (keszybz) [reply units/systemd-boot-update.service]: {{ROOTBINDIR}} ain't gonna work either, this is not an .in file ;P

Please just just drop the path fully.

**Comment by keszybz** [comment]
> It's now enabled via presets and via build scripts. Does this even matter for Fedora?

Yes, it matters in the sense that both thing would need to be "undone" in Fedora. The preset edited from the file and the symlink removed from the filesystem.

There certainly shouldn't be the static installation symlink if the unit has installation instructions. 

If the preset is "on" by default, we'll just revert that in Fedora. I think that's the appropriate choice for binary package distros… There is really no expectation of updates being interrupted halfway. And even if something like that were to happen, the user will have to redo the transaction anyway, so the update will happen as expected.

> Also, isn't the sd-boot EFI binary its own RPM anyway on Fedora? If so this unit file should be shipped in that I figure.

So far not.

File: units/systemd-boot-update.service (hunk 1 above)
Comment (poettering) [reply units/systemd-boot-update.service]: relying on PATH for executing binaries we invoke is still a really bad idea, I still think we shouldn't have changed our code to use that.

**Comment by poettering** [comment]
> > It's now enabled via presets and via build scripts. Does this even matter for Fedora?
> 
> Yes, it matters in the sense that both thing would need to be "undone" in Fedora. The preset edited from the file and the symlink removed from the filesystem.
> 
> There certainly shouldn't be the static installation symlink if the unit has installation instructions.
> 
> If the preset is "on" by default, we'll just revert that in Fedora. I think that's the appropriate choice for binary package distros… There is really no expectation of updates being interrupted halfway. And even if something like that were to happen, the user will have to redo the transaction anyway, so the update will happen as expected.

So, I wouldn't see "binary package distros" and "image based distros" as two opposing concepts. My own focus is on building an image based OS off a binary package distro, so that one can take benefit of RPMs/DEBs but only use them for building the images on the vendor side, and then deploy things in an image-based way. For cases like that I certainly want the bootctl auto logic.

I still think we should enable this by default via presets from upstream if the boot loader stuff is enabled at all, because it's carefully written to be a NOP when there's nothing to update. And I'd really like to have the testing/feedback on that. i.e. it really shouldn't hurt to leave this on, and I really want to know about failures with it.

I mean, if Fedora doesn#t want to be the testbed for this, then that's OK, but I still think we should make this the upstream default, so that we can collect the feedback wherever people are leaving things on.

File: units/systemd-boot-update.service (hunk 1 above)
Comment (mbiebl) [reply units/systemd-boot-update.service]: > relying on PATH for executing binaries we invoke is still a really bad idea

I think it makes sense to rely on PATH for *external* tools, but for our internal binaries using the full path is preferrable.

**Comment by poettering** [comment]
force pushed a new version now. only change is the path fix. I hope this passes CI now.

File: units/meson.build [hunk 2]
```diff
@@ -104,6 +104,8 @@ units = [
         ['systemd-ask-password-wall.service',   ''],
         ['systemd-boot-system-token.service',   'ENABLE_EFI',
          'sysinit.target.wants/'],
+        ['systemd-boot-update.service',         'ENABLE_EFI HAVE_BLKID',
+         'sysinit.target.wants/'],
```
Comment (keszybz) [inline units/meson.build]: … please drop this line.

**Review by keszybz — COMMENTED** [review]
> My own focus is on building an image based OS off a binary package distro, so that one can take benefit of RPMs/DEBs but only use them for building the images on the vendor side, and then deploy things in an image-based way. For cases like that I certainly want the bootctl auto logic.

Yeah, that is a very good argument.

> I still think we should enable this by default via presets from upstream if the boot loader stuff is enabled at all, because it's carefully written to be a NOP when there's nothing to update.

Makes sense.



> There certainly shouldn't be the static installation symlink if the unit has installation instructions.

I guess I wasn't clear there: right now the unit is both statically enabled by a symlink in `/usr/lib/systemd/system/sysinit.target.wants/` that is created at installation time, but could also be enabled by `systemctl preset` or `systemctl enable`. We only want this second mechanism, i.e. …

**Comment by poettering** [comment]
force pushed a new version, only dropping the suggested line. Since there were no other complaints taking liberty to set green label

**Pull request merged**